src/lib/tuning/greedy_selector.cpp

```diff
@@ -0,0 +1,143 @@
+  // Assumption: cost() >= 0 ==> accept_cost() >= 0 && current_cost() >= 0 && reject_cost() <= 0
+
+  // Accumulate absolute cost balance from currently chosen choices.
+  float cost_balance = 0.0f;
+  for (const auto& choice : choices) {
+    cost_balance += choice->current_cost();
+  }
+
+  // Desirability balance is always relative to current system state.
+  float desirability_balance = 0.0f;
+
+  // Sort choices:
+  //  1) by reject_cost() ascending
+  //  2) by accept_desirability() ascending stable
+  // The most desirable choice is at back() and among equaly desirable choices
+  // the least costly choices to accept are nearer back().
+  std::list<std::shared_ptr<TuningChoice>> sorted_choices(choices.size());
+  std::copy(choices.cbegin(), choices.cend(), sorted_choices.begin());
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->reject_cost() < rhs->reject_cost();
+  });
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->accept_desirability() < rhs->accept_desirability();
+  });
+
+  // If current state exceeds cost_budget,
+  // reject the least desirable choices to reduce cost_balance.
+  if (cost_balance > cost_budget) {
+    LOG_INFO("Cost balance of " << cost_balance << " exceeds budget of " << cost_budget);
+    auto sacrifice_until = sacrifice_choices(sorted_choices, cost_budget - cost_balance);
+    if (sacrifice_until == sorted_choices.end()) {
+      LOG_WARN("Cost budget is impossible to maintain. No Operations are performed.");
+      return operations;
+    }
+    for (auto choice = sorted_choices.begin(); choice != sacrifice_until; ++choice) {
+      LOG_DEBUG(" ! Reject " << **choice << " to reduce cost balance");
+      operations.push_back((*choice)->reject());
+      cost_balance += (*choice)->reject_cost();
+      desirability_balance += (*choice)->reject_desirability();
+    }
+    sorted_choices.erase(sorted_choices.begin(), sacrifice_until);
+  }
+
+  // Select the most desirable operations first
+  // always maintaining the cost budget
+  while (sorted_choices.size() > 0) {
+    if (sorted_choices.front()->reject_desirability() > sorted_choices.back()->accept_desirability()) {
+      LOG_DEBUG("Rejecting " << *sorted_choices.front() << " is most beneficial.");
+      // Rejecting a choice can only reduce cost_balance and never exceed cost_budget (as cost()>=0)
+
+      LOG_DEBUG(" ! Reject " << *sorted_choices.front());
+      operations.push_back(sorted_choices.front()->reject());
+      cost_balance += sorted_choices.front()->reject_cost();
+      desirability_balance += sorted_choices.front()->reject_desirability();
+      sorted_choices.pop_front();
+
+    } else {
+      LOG_DEBUG("Accepting " << *sorted_choices.back() << " is most beneficial.");
+      // Accepting a choice could exceed cost_budget, so reduce cost_balance first
+
+      auto sacrifice_until =
```

Comment:
Could you add a comment here?

Reply:
There is already a comment a few lines up and there is documentation on the `sacrifice_choices` function, but maybe it's not enough. What exactly is unclear or how could we make this better?